File: 백준/Silver/2210. 숫자판 점프/숫자판 점프.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");
const board = input.map((line) => line.split(" "));

const uniqueNumbers = new Set();

const directions = [
  [-1, 0],
  [1, 0],
  [0, -1],
  [0, 1],
];

function dfs(x, y, currentNumber) {
  if (currentNumber.length === 6) {
    uniqueNumbers.add(currentNumber);
    return;
  }

  for (const [dx, dy] of directions) {
    const nx = x + dx;
    const ny = y + dy;

    if (nx >= 0 && nx < 5 && ny >= 0 && ny < 5) {
      dfs(nx, ny, currentNumber + board[nx][ny]);
    }
  }
}

for (let i = 0; i < 5; i++) {
  for (let j = 0; j < 5; j++) {
    dfs(i, j, board[i][j]);
  }
}

console.log(uniqueNumbers.size);

File: 백준/Silver/2210. 숫자판 점프/숫자판 점프.py
def dfs(x, y, number):
    if len(number) == 6:
        unique_numbers.add(number)
        return
    
    for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
        nx, ny = x + dx, y + dy
        if 0 <= nx < 5 and 0 <= ny < 5:  # 격자 범위를 벗어나지 않으면
            dfs(nx, ny, number + board[nx][ny])

board = [input().split() for _ in range(5)]
unique_numbers = set()

for i in range(5):
    for j in range(5):
        dfs(i, j, board[i][j])

print(len(unique_numbers))